File: smart-table/src/db/services/dashboardShareService.ts
import { db } from "../schema";
import type { DashboardShare } from "../schema";
import { generateId } from "../../utils/id";
import { apiClient } from "@/api/client";

export interface CreateShareData {
  dashboardId: string;
  expiresInHours?: number;
  maxAccessCount?: number;
  requireAccessCode?: boolean;
  permission?: "view" | "edit";
}

export interface ShareValidationResult {
  valid: boolean;
  share?: DashboardShare;
  error?: string;
}

export class DashboardShareService {
  /**
   * 生成随机分享令牌
   */
  private generateShareToken(): string {
    const chars =
      "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
    let token = "";
    for (let i = 0; i < 16; i++) {
      token += chars.charAt(Math.floor(Math.random() * chars.length));
    }
    return token;
  }

  /**
   * 生成访问密码
   */
  private generateAccessCode(): string {
    const digits = "0123456789";
    let code = "";
    for (let i = 0; i < 6; i++) {
      code += digits.charAt(Math.floor(Math.random() * digits.length));
    }
    return code;
  }

  /**
   * 创建分享链接 - 调用后端 API
   */
  async createShare(data: CreateShareData): Promise<DashboardShare> {
    try {
      console.log("[DashboardShare] Creating share:", data);

      // 调用后端 API 创建分享
      const response = await apiClient.post(
        `/dashboards/${data.dashboardId}/shares`,
        {
          requireAccessCode: data.requireAccessCode,
          expiresInHours: data.expiresInHours,
          maxAccessCount: data.maxAccessCount,
          permission: data.permission || "view",
        },
      );

      console.log("[DashboardShare] API response:", response);

      // 适配后端响应格式
      // 可能是 {success, message, data} 或直接返回数据对象
      let apiShare: any;
      if (response.data?.data) {
        apiShare = response.data.data;
      } else if (response.data) {
        apiShare = response.data;
      } else {
        apiShare = response;
      }

      console.log("[DashboardShare] Extracted share:", apiShare);

      if (!apiShare || !apiShare.id) {
        throw new Error("Invalid API response: missing share data");
      }

      // 转换为本地格式并保存到 IndexedDB
      const share: DashboardShare = {
        id: apiShare.id,
        dashboardId: apiShare.dashboard_id,
        shareToken: apiShare.share_token,
        accessCode: apiShare.has_access_code
          ? this.generateAccessCode()
          : undefined,
        expiresAt: apiShare.expires_at,
        maxAccessCount: apiShare.max_access_count,
        currentAccessCount: apiShare.current_access_count,
        isActive: apiShare.is_active,
        permission: apiShare.permission as "view" | "edit",
        createdAt: new Date(apiShare.created_at).getTime(),
        createdBy: apiShare.created_by,
      };

      await db.dashboardShares.put(share);
      return share;
    } catch (error) {
      console.error("[DashboardShare] createShare failed:", error);
      throw error;
    }
  }

  /**
   * 通过分享令牌获取分享信息 - 调用后端 API（公开接口）
   */
  async getShareByToken(token: string): Promise<DashboardShare | undefined> {
    try {
      // 先尝试从本地获取
      const localShare = await db.dashboardShares
        .where("shareToken")
        .equals(token)
        .first();
      if (localShare) {
        return localShare;
      }

      // 本地没有，从后端获取（这个接口需要 accessCode）
      // 由于是公开接口，我们只从本地读取
      return undefined;
    } catch (error) {
      console.error("[DashboardShare] getShareByToken failed:", error);
      return undefined;
    }
  }

  /**
   * 获取仪表盘的所有分享链接 - 调用后端 API
   */
  async getSharesByDashboard(dashboardId: string): Promise<DashboardShare[]> {
    try {
      console.log(
        "[DashboardShare] Fetching shares from API for dashboard:",
        dashboardId,
      );

      // 检查是否有 token
      const token =
        localStorage.getItem("access_token") ||
        sessionStorage.getItem("access_token");
      console.log(
        "[DashboardShare] Current token:",
        token ? "exists" : "missing",
      );

      // 调用后端 API 获取分享列表
      const response = await apiClient.get(`/dashboards/${dashboardId}/shares`);
      console.log("[DashboardShare] API response:", response);

      // 适配后端响应格式 {success, message, data}
      const apiShares = response.data?.data || response.data || [];
      console.log("[DashboardShare] Extracted shares:", apiShares);

      if (!Array.isArray(apiShares)) {
        console.warn(
          "[DashboardShare] API response is not an array:",
          apiShares,
        );
        return [];
      }

      // 转换为本地格式并保存
      const shares: DashboardShare[] = await Promise.all(
        apiShares.map(async (apiShare: any) => {
          const share: DashboardShare = {
            id: apiShare.id,
            dashboardId: apiShare.dashboard_id,
            shareToken: apiShare.share_token,
            accessCode: apiShare.has_access_code
              ? this.generateAccessCode()
              : undefined,
            expiresAt: apiShare.expires_at,
            maxAccessCount: apiShare.max_access_count,
            currentAccessCount: apiShare.current_access_count,
            isActive: apiShare.is_active,
            permission: apiShare.permission as "view" | "edit",
            createdAt: new Date(apiShare.created_at).getTime(),
            createdBy: apiShare.created_by,
          };
          await db.dashboardShares.put(share);
          return share;
        }),
      );

      console.log("[DashboardShare] Loaded", shares.length, "shares from API");
      return shares;
    } catch (error: any) {
      console.error("[DashboardShare] getSharesByDashboard failed:", error);
      console.error("[DashboardShare] Error details:", {
        message: error.message,
        response: error.response?.data,
        status: error.response?.status,
        headers: error.response?.headers,
      });

      // API 失败时从本地读取
      console.warn("[DashboardShare] Falling back to local cache");
      const shares = await db.dashboardShares
        .where("dashboardId")
        .equals(dashboardId)
        .and((share) => share.isActive)
        .sortBy("createdAt");
      return shares;
    }
  }

  /**
   * 验证分享链接是否有效 - 调用后端 API（公开接口）
   */
  async validateShare(
    token: string,
    accessCode?: string,
  ): Promise<ShareValidationResult> {
    try {
      console.log("[DashboardShare] Validating share:", token);

      // 调用后端 API 验证
      const response = await apiClient.post(`/shares/${token}/validate`, {
        accessCode,
      });

      console.log("[DashboardShare] Full response:", response);

      // 适配多种响应格式
      let result: any;

      // 情况 1: Axios 标准格式 {data: {success, message, data: {share, dashboard}}}
      if (response.data?.data?.share && response.data?.data?.dashboard) {
        result = response.data.data;
      }
      // 情况 2: 直接返回数据 {data: {share, dashboard}}
      else if (response.data?.share && response.data?.dashboard) {
        result = response.data;
      }
      // 情况 3: response 本身就是 {share, dashboard}
      else if (response.share && response.dashboard) {
        result = response;
      }
      // 情况 4: 其他格式
      else {
        result = response.data || response;
      }

      console.log("[DashboardShare] Extracted result:", result);

      if (!result) {
        return {
          valid: false,
          error: "验证失败：无响应数据",
        };
      }

      if (result.share && result.dashboard) {
        const { share, dashboard, tables } = result;

        // 保存 share 到本地
        const localShare: DashboardShare = {
          id: share.id,
          dashboardId: share.dashboard_id,
          shareToken: share.share_token,
          accessCode: accessCode,
          expiresAt: share.expires_at,
          maxAccessCount: share.max_access_count,
          currentAccessCount: share.current_access_count,
          isActive: share.is_active,
          permission: share.permission as "view" | "edit",
          createdAt: new Date(share.created_at).getTime(),
          createdBy: share.created_by,
        };
        await db.dashboardShares.put(localShare);

        // 保存 dashboard 到本地
        const localDashboard: Dashboard = {
          id: dashboard.id,
          baseId: dashboard.base_id,
          name: dashboard.name,
          description: dashboard.description,
          widgets: (dashboard.widgets as any[]) || [],
          layout: dashboard.layout || {},
          layoutType:
            ((dashboard.layout as any)?.type as "grid" | "free") || "grid",
          gridColumns: (dashboard.layout as any)?.config?.gridColumns || 12,
          refreshConfig: {
            enabled: false,
            interval: 30000,
            autoRefresh: false,
          },
          isStarred: false,
          order: 0,
          createdAt: new Date(dashboard.created_at).getTime(),
          updatedAt: new Date(dashboard.updated_at).getTime(),
        };
        await db.dashboards.put(localDashboard);

        // 保存表数据到本地（如果后端返回了 tables）
        if (tables && Array.isArray(tables)) {
          console.log(
            "[DashboardShare] Saving",
            tables.length,
            "tables to local cache",
          );
          for (const table of tables) {
            // 保存表信息到 tableEntities 表
            await db.tableEntities.put({
              id: table.id,
              baseId: table.base_id,
              name: table.name,
              description: table.description,
              order: table.order || 0,
              primaryFieldId: table.primary_field_id,
              createdAt: new Date(table.created_at).getTime(),
              updatedAt: new Date(table.updated_at).getTime(),
            });
            console.log(
              `  [DashboardShare] 保存表：${table.name} (${table.id})`,
            );

            // 保存字段信息
            if (table.fields && Array.isArray(table.fields)) {
              console.log(`    保存 ${table.fields.length} 个字段`);
              for (const field of table.fields) {
                await db.fields.put({
                  id: field.id,
                  tableId: field.table_id,
                  name: field.name,
                  type: field.type,
                  description: field.description,
                  order: field.order || 0,
                  isPrimary: field.is_primary || false,
                  isRequired: field.is_required || false,
                  options: field.options || null,
                  config: field.config || null,
                  createdAt: new Date(field.created_at).getTime(),
                  updatedAt: new Date(field.updated_at).getTime(),
                });
              }
            }

            // 保存记录信息
            if (table.records && Array.isArray(table.records)) {
              console.log(
                `  Table ${table.name}: saving ${table.records.length} records`,
              );
              for (const record of table.records) {
                await db.records.put({
                  id: record.id,
                  tableId: record.table_id,
                  values: record.values || {},
                  created_at: new Date(record.created_at).getTime(),
                  updated_at: new Date(record.updated_at).getTime(),
                });
              }
            }
          }

          // 验证数据是否真的被保存了
          const savedTables = await db.tableEntities.toArray();
          const savedFields = await db.fields.toArray();
          const savedRecords = await db.records.toArray();
          console.log(
            `[DashboardShare] 验证保存结果：${savedTables.length} 个表，${savedFields.length} 个字段，${savedRecords.length} 条记录`,
          );
        }

        return {
          valid: true,
          share: localShare,
          dashboard: localDashboard,
          tables: tables || [], // 返回表数据
        };
      } else {
        console.warn(
          "[DashboardShare] Invalid result format, expected share and dashboard:",
          result,
        );
        return {
          valid: false,
          error: "验证失败：数据格式错误",
        };
      }
    } catch (error: any) {
      console.error("[DashboardShare] validateShare failed:", error);
      const errorMsg =
        error.response?.data?.message || error.message || "验证失败";
      return {
        valid: false,
        error: errorMsg,
      };
    }
  }

  /**
   * 记录访问 - 调用后端 API
   */
  async recordAccess(shareId: string): Promise<void> {
    try {
      // 后端在验证时自动记录，这里只需要更新本地
      const share = await db.dashboardShares.get(shareId);
      if (!share) return;

      await db.dashboardShares.update(shareId, {
        currentAccessCount: share.currentAccessCount + 1,
        lastAccessedAt: Date.now(),
      });
    } catch (error) {
      console.error("[DashboardShare] recordAccess failed:", error);
    }
  }

  /**
   * 禁用分享链接 - 调用后端 API
   */
  async deactivateShare(shareId: string): Promise<void> {
    try {
      // 调用后端 API 禁用
      await apiClient.post(`/shares/${shareId}/deactivate`);

      // 更新本地
      await db.dashboardShares.update(shareId, {
        isActive: false,
      });
    } catch (error) {
      console.error("[DashboardShare] deactivateShare failed:", error);
      throw error;
    }
  }

  /**
   * 删除分享链接 - 调用后端 API
   */
  async deleteShare(shareId: string): Promise<void> {
    try {
      // 调用后端 API 删除
      await apiClient.delete(`/shares/${shareId}`);

      // 从本地删除
      await db.dashboardShares.delete(shareId);
    } catch (error) {
      console.error("[DashboardShare] deleteShare failed:", error);
      throw error;
    }
  }

  /**
   * 清理过期的分享链接
   */
  async cleanupExpiredShares(): Promise<number> {
    const now = Date.now();
    const expiredShares = await db.dashboardShares
      .where("expiresAt")
      .below(now)
      .and((share) => share.isActive)
      .toArray();

    for (const share of expiredShares) {
      await db.dashboardShares.update(share.id, { isActive: false });
    }

    return expiredShares.length;
  }

  /**
   * 生成分享链接 URL
   */
  generateShareUrl(token: string): string {
    const baseUrl = window.location.origin;
    return `${baseUrl}/#/share/dashboard/${token}`;
  }

  /**
   * 复制链接到剪贴板
   */
  async copyToClipboard(text: string): Promise<boolean> {
    try {
      if (navigator.clipboard && window.isSecureContext) {
        await navigator.clipboard.writeText(text);
        return true;
      }

      // 降级方案
      const textArea = document.createElement("textarea");
      textArea.value = text;
      textArea.style.position = "fixed";
      textArea.style.left = "-999999px";
      textArea.style.top = "-999999px";
      document.body.appendChild(textArea);
      textArea.focus();
      textArea.select();

      const result = document.execCommand("copy");
      document.body.removeChild(textArea);
      return result;
    } catch (error) {
      console.error("复制失败:", error);
      return false;
    }
  }
}

export const dashboardShareService = new DashboardShareService();

File: smarttable-backend/app/services/dashboard_share_service.py
"""
仪表盘分享服务模块
"""
from typing import List, Optional, Dict, Any
import uuid
import secrets
import string
from datetime import datetime, timezone, timedelta

from flask import current_app
from app.extensions import db, cache
from app.models.dashboard_share import DashboardShare, SharePermission


# 密码尝试限制配置
ACCESS_CODE_MAX_ATTEMPTS = 5       # 最大尝试次数
ACCESS_CODE_LOCKOUT_SECONDS = 900   # 锁定时间（15 分钟）


class DashboardShareService:
    """仪表盘分享服务类"""
    
    @staticmethod
    def generate_share_token() -> str:
        """生成随机分享令牌"""
        return secrets.token_urlsafe(32)
    
    @staticmethod
    def generate_access_code() -> str:
        """生成 6 位数字访问密码"""
        return ''.join(secrets.choice(string.digits) for _ in range(6))
    
    @staticmethod
    def create_share(
        dashboard_id: str,
        user_id: str,
        require_access_code: bool = False,
        expires_in_hours: Optional[int] = None,
        max_access_count: Optional[int] = None,
        permission: str = 'view',
        title: Optional[str] = None,
        description: Optional[str] = None
    ) -> DashboardShare:
        """
        创建分享链接
        
        参数:
            dashboard_id: 仪表盘 ID
            user_id: 创建用户 ID
            require_access_code: 是否需要访问密码
            expires_in_hours: 过期时间（小时）
            max_access_count: 最大访问次数
            permission: 分享权限（view/edit）
            title: 分享标题
            description: 分享备注
            
        返回:
            创建的分享对象
        """
        # 生成分享令牌
        share_token = DashboardShareService.generate_share_token()
        
        # 生成访问密码（如果需要）
        access_code = DashboardShareService.generate_access_code() if require_access_code else None
        
        # 计算过期时间
        expires_at = None
        if expires_in_hours:
            expires_at = int((datetime.now(timezone.utc) + timedelta(hours=expires_in_hours)).timestamp())
        
        # 创建分享对象
        share = DashboardShare(
            dashboard_id=dashboard_id,
            share_token=share_token,
            title=title,
            description=description,
            access_code=access_code,
            expires_at=expires_at,
            max_access_count=max_access_count,
            current_access_count=0,
            is_active=True,
            permission=SharePermission(permission),
            created_by=user_id
        )
        
        db.session.add(share)
        db.session.commit()
        
        return share
    
    @staticmethod
    def get_share_by_token(token: str) -> Optional[DashboardShare]:
        """
        通过分享令牌获取分享信息
        
        参数:
            token: 分享令牌
            
        返回:
            分享对象或 None
        """
        return DashboardShare.query.filter_by(share_token=token).first()
    
    @staticmethod
    def get_shares_by_dashboard(dashboard_id: str) -> List[DashboardShare]:
        """
        获取仪表盘的所有分享链接（含已停用的）
        
        参数:
            dashboard_id: 仪表盘 ID
            
        返回:
            分享列表
        """
        return DashboardShare.query.filter_by(
            dashboard_id=dashboard_id
        ).order_by(DashboardShare.created_at.desc()).all()
    
    @staticmethod
    def validate_share(
        token: str,
        access_code: Optional[str] = None,
        client_ip: Optional[str] = None
    ) -> tuple[bool, Optional[DashboardShare], Optional[str]]:
        """
        验证分享链接是否有效
        
        参数:
            token: 分享令牌
            access_code: 访问密码
            client_ip: 客户端 IP（用于密码尝试限制）
            
        返回:
            (是否有效，分享对象，错误码)
            
        错误码:
            share_not_found - 分享不存在
            share_deactivated - 分享已禁用
            share_expired - 分享已过期
            share_access_limit - 访问次数达上限
            requires_access_code - 需要访问密码
            access_code_locked - 密码尝试次数过多
            access_code_invalid - 访问密码错误
        """
        share = DashboardShareService.get_share_by_token(token)
        
        if not share:
            return False, None, 'share_not_found'
        
        if not share.is_active:
            return False, share, 'share_deactivated'
        
        # 检查是否过期
        if share.expires_at and datetime.now(timezone.utc).timestamp() > share.expires_at:
            return False, share, 'share_expired'
        
        # 检查访问次数
        if share.max_access_count and share.current_access_count >= share.max_access_count:
            return False, share, 'share_access_limit'
        
        # 验证访问密码
        if share.access_code:
            if access_code is None:
                # 未提供访问密码，但分享需要密码
                return False, share, 'requires_access_code'
            
            # 检查密码尝试限制
            lockout_key = f"share_access_code_lockout:{token}:{client_ip or 'unknown'}"
            attempts_key = f"share_access_code_attempts:{token}:{client_ip or 'unknown'}"
            
            if cache.get(lockout_key):
                return False, share, 'access_code_locked'
            
            # 使用时间安全比较防止时序攻击
            if not secrets.compare_digest(share.access_code, access_code):
                # 增加失败计数
                attempts = (cache.get(attempts_key) or 0) + 1
                cache.set(attempts_key, attempts, timeout=ACCESS_CODE_LOCKOUT_SECONDS)
                
                if attempts >= ACCESS_CODE_MAX_ATTEMPTS:
                    # 锁定并清除尝试计数
                    cache.set(lockout_key, True, timeout=ACCESS_CODE_LOCKOUT_SECONDS)
                    cache.delete(attempts_key)
                    return False, share, 'access_code_locked'
                
                return False, share, 'access_code_invalid'
            
            # 密码正确，清除失败计数
            cache.delete(attempts_key)
            cache.delete(lockout_key)
        
        return True, share, None
    
    @staticmethod
    def record_access(
        share_id: str,
        client_ip: Optional[str] = None,
        user_agent: Optional[str] = None
    ) -> bool:
        """
        记录访问
        
        参数:
            share_id: 分享 ID
            client_ip: 客户端 IP
            user_agent: 客户端 User-Agent
            
        返回:
            是否成功
        """
        share = DashboardShare.query.get(share_id)
        if not share:
            return False
        
        share.current_access_count += 1
        share.last_accessed_at = datetime.now(timezone.utc)
        db.session.commit()
        
        # 记录访问日志
        current_app.logger.info(
            f'[ShareAccess] share_id={share_id} token={share.share_token[:8]}... '
            f'ip={client_ip or "unknown"} ua={user_agent or "unknown"} '
            f'total_access={share.current_access_count}'
        )
        
        return True
    
    @staticmethod
    def deactivate_share(share_id: str) -> bool:
        """
        禁用分享链接
        
        参数:
            share_id: 分享 ID
            
        返回:
            是否成功
        """
        share = DashboardShare.query.get(share_id)
        if not share:
            return False
        
        share.is_active = False
        db.session.commit()
        
        return True
    
    @staticmethod
    def delete_share(share_id: str) -> bool:
        """
        删除分享链接
        
        参数:
            share_id: 分享 ID
            
        返回:
            是否成功
        """
        share = DashboardShare.query.get(share_id)
        if not share:
            return False
        
        db.session.delete(share)
        db.session.commit()
        
        return True
    
    @staticmethod
    def cleanup_expired_shares() -> int:
        """
        清理过期的分享链接
        
        返回:
            清理的数量
        """
        now = int(datetime.now(timezone.utc).timestamp())
        expired_shares = DashboardShare.query.filter(
            DashboardShare.is_active == True,
            DashboardShare.expires_at != None,
            DashboardShare.expires_at < now
        ).all()
        
        count = 0
        for share in expired_shares:
            share.is_active = False
            count += 1
        
        if count > 0:
            db.session.commit()
        
        return count
